Unit tests in venia can run against stale peregrine code

<!-- (REQUIRED) What is the nature of this issue? -->

## This issue is for the following packages:

[X] `venia-concept`
[ ] `pwa-buildpack`
[ ] `peregrine`
[ ] `pwa-module`

## This issue is a:

[X] Bug
[ ] Feature suggestion
[ ] Other (Please Specify)

### Environment

| Question                    |    Answer     |
| --------------------------- | :-----------: |
| Magento version             | <!--Answer--> |
| Operating System + version  | <!--Answer--> |
| node.js version (`node -v`) | <!--Answer--> |
| npm version (`npm -v`)      | <!--Answer--> |

## Description
**Steps to Replicate**
1. Run `npm install` in the root of the repository to bootstrap
2. Run `rm -rf packages/peregrine/dist` in the root of the repository
3. Run `npm test` in the root of the repository

Tests in `venia-concept` should fail with `Cannot find module '@magento/peregrine' from 'foobar.js'`. This means tests in `venia-concept` that consume `@magento/peregrine` are pulling in the last built version of the lib locally, which may not be up-to-date with the `src` files in `@magento/peregrine`. 
<!-- (REQUIRED) Describe the issue with as much detail as possible -->

## Expected result:
Cross-package unit/integration tests always consume from `src`, not `dist`. The `dist` artifact should only ever be used for publishing to `npm`, and nothing else.

<!-- (REQUIRED) When this issue is resolved, what is the expected result/behavior? -->
Cross-package unit/integration tests always consume from `src`, not `dist`.
## Possible solutions:
Use Jest's configuration to re-write paths for libs in this monorepo, to ensure that:

1. Jest handles transpiling the `src` of the libs in this repo with Babel
2. Jest pulls files from `src`, rather than the `dist` it finds in the package's `main` entry in `package.json`.
<!-- (OPTIONAL) What would a solution for this issue look like? -->

<!--
Thank you for taking the time to report this issue!
GitHub Issues should only be created for problems/topics related to this project's codebase.

Before submitting this issue, please make sure you are complying with our Code of Conduct:
https://github.com/magento-research/venia-pwa-concept/blob/develop/.github/CODE_OF_CONDUCT.md

Issues that do not comply with our Code of Conduct or do not contain enough information may be closed at the maintainers' discretion.

Feel free to remove this section before creating this issue.
-->

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -13,7 +13,10 @@ jobs:
           - v1-npm-cache-{{ .Branch }}
           - v1-npm-cache-
 
-      # Install root devDependencies and bootstrap packages
+      # Install root devDependencies and bootstrap packages. We use
+      # --ignore-scripts to prevent building any artifacts to disk prior
+      # to running Jest. This is a safeguard to ensure our tests always operate
+      # on the latest source
       - run:
           name: Install shared devDependencies
           command: 'npm install --ignore-scripts'

--- a/packages/venia-concept/jest.config.js
+++ b/packages/venia-concept/jest.config.js
@@ -1,8 +1,19 @@
+const { join } = require('path');
+
 module.exports = {
     displayName: 'Venia Concept',
     browser: true,
     moduleNameMapper: {
         '\\.css$': 'identity-obj-proxy',
-        '^src/(.+)': '<rootDir>/src/$1'
-    }
+        // Mirrors webpack alias to resolve from 'src'
+        '^src/(.+)': '<rootDir>/src/$1',
+        // Re-write imports to Peregrine to ensure they're not pulled from the
+        // (possibly outdated) build artifacts on disk in `dist`.
+        // Ideally this rule would be in the root Jest config, but Jest's config
+        // merging strategy isn't currently smart enough for this
+        '^@magento/peregrine(/*(?:.+)*)': join(__dirname, '../peregrine/src/$1')
+    },
+    // Have Jest use Babel to transpile Peregrine imports in tests, since
+    // our cross-package tests in the monorepo should all operate on `src`
+    transformIgnorePatterns: ['node_modules/(?!@magento/peregrine)']
 };

--- a/packages/venia-concept/jest.config.js
+++ b/packages/venia-concept/jest.config.js
@@ -10,7 +10,9 @@ module.exports = {
         // Re-write imports to Peregrine to ensure they're not pulled from the
         // (possibly outdated) build artifacts on disk in `dist`.
         // Ideally this rule would be in the root Jest config, but Jest's config
-        // merging strategy isn't currently smart enough for this
+        // merging strategy isn't currently smart enough for this. TODO: Look
+        // into moving to root config in later versions of Jest if config merging
+        // improves
         '^@magento/peregrine(/*(?:.+)*)': join(__dirname, '../peregrine/src/$1')
     },
     // Have Jest use Babel to transpile Peregrine imports in tests, since

--- a/packages/venia-concept/package.json
+++ b/packages/venia-concept/package.json
@@ -16,7 +16,7 @@
     "start:debug": "node --inspect-brk ./node_modules/.bin/webpack-dev-server --progress --color --env.phase development"
   },
   "dependencies": {
-    "@magento/peregrine": "^0.3.0",
+    "@magento/peregrine": "^0.4.0",
     "babel-runtime": "^6.26.0",
     "dotenv": "^4.0.0",
     "feather-icons": "^4.6.0",

